### ISSUE TITLE
Question: Optimization and Dask dataframe to_parquet

### ISSUE THREAD

**Issue by hoangthienan95** [dask/dask#4795]
Hello,
I'm trying to both use Dask with my workflow and experiment with different ways to do the same task to learn Dask and get an intuition of which API (array, dataframe, bag, delayed) to use. Therefore, some feedback on the thought process would be greatly appreciated. I hope that I can learn intuition/principles from a small example and extrapolate to bigger workflows. I also have questions about how Dask dataframe is written to parquet.

Below is a subset of my computational workflow :
![image](https://user-images.githubusercontent.com/25307953/57666322-92041f80-75cd-11e9-93dc-0a6524f5830c.png)

Some background: the final Dask Dataframe can have up to 100k columns and 1 million rows. The parser for my input data returns `test_arr` as Dask-delayed objects so I can't change that. I need each each `test_arr` to map to a column because in next step of the pipeline, I only need a subset of columns as any given moment (but all the columns should be batched together as a dataset entity).

I'm only showing 2 conditions and 2 choices in `np.select`, but I have up to 10 pairs with very different logic.

Is there a better way to do what I did? I resorted to using `delayed` all the way for my use case because: (these opinions are from my limited experience, please correct me if I'm wrong)

1. *Dask Dataframe*: hard to initialize efficiently/ if you are not reading it from a file (`csv`,`parquet` etc). There is no easy way to initialize dataframe by columns, only through [assigning one by one or `dd.concat`](https://stackoverflow.com/questions/48090036/create-dask-dataframe-from-a-list-of-dask-series)

1. *Dask Array* : Similar to `Dataframe`. Since there is no `dask.array.select`to mimic the numpy API, I would have to use numpy then convert to dask array, and doing so would create a huge array in memory, which defeats the purpose. I couldn't convert my `test_arr` to a `Dask array` using `from_delayed` because I need to know the rows for the `shape` argument in `da.from_delayed`, which results in computing the array to memory. Even then, the lack of the convenience function `da.select` would make my code more lengthy than needed due to combination of multiple conditions and choices.

1. *Dask bag*: Although Dask bag can be used in much more flexible ways, it suffers in speed. I have tried mapping simple functions onto dask bag and realize that it is much much slower. Even when I have a sequence of length in the millions, I would [batch them with delayed](http://docs.dask.org/en/latest/delayed-best-practices.html#avoid-too-many-tasks) instead of using bags. In my mind, bags are the last resort when nothing else works. Also, I'm doing mostly numpy/pandas calls so I would lose out on performance with `bag`. Also, to convert `bag` into `dataframe`,  each element in the bag should be a row, whereas I have each entry as a column.

From the above, the logical way is to use `delayed`, flexible, fast, and take advantage of already-exist numpy functions.

Parquet: With a table of this size (100k column and 1 million rows), I'm afraid I can't pull more than one column of the data in memory of local cluster at any point, up to outputting to parquet (talking to IT at my work to lower their restrictions on our cluster to allow `dask-worker` jobs). When outputting to parquet, does Dask do it one column at a time? How can I specify that each column is an independent unit?

If there are some resources on any of my questions above,  a link instead of explanation is more than enough. I might have missed something while reading the documentation. Thanks in advance!

**Comment by hoangthienan95** [comment]
Following up on this, I realize that my dataframe had only one partition. However, I have no idea how to repartition it. This becomes an issue when I try to construct the final dataframe:

With this df as the basic structure
![image](https://user-images.githubusercontent.com/25307953/57728181-456a2400-7661-11e9-9273-bba2f35625b5.png)

I tried to append those delayed columns but encountering trouble:
```python
for col_name, col in tqdm(rsid_col_tuple_list): #rsid_col_tuple_list:[(name, dask_delayed(np.select))]
    final_df[col_name] = dd.from_delayed(col, meta=(col_name, "<U5")) 
```
raises `Not all divisions are known, can't align partitions. Please use `set_index` to set the index` 

when I add `.compute()` it works just fine.
```python
for col_name, col in tqdm(rsid_col_tuple_list): #rsid_col_tuple_list:[(name, dask_delayed(np.select))]
    final_df[col_name] = dd.from_delayed(col, meta=(col_name, "<U5")).compute()
```
I understand that it is because the divisions are mismatched between the columns to be added and the dataframe, but can't find a way to generate/insert that division metadata into the column. Doing `dd.from_delayed(col, meta=(col_name, "<U5"), divisions = final_df.divisions)` didn't work.

I think this should be easy and I'm just missing some key things right now. I have looked at many sources: ([1](https://github.com/dask/dask/issues/1426), [2](https://stackoverflow.com/questions/48158272/add-a-dask-array-column-to-a-dask-dataframe),[3](https://stackoverflow.com/questions/46923274/appending-new-column-to-dask-dataframe/46951629#46951629),[4](https://stackoverflow.com/questions/46911220/simple-way-to-dask-concatenate-horizontal-axis-1-columns)) but still can't figure it out. I'd really appreciate any pointers in the right direction.

**Comment by TomAugspurger** [comment]
It'd be a bit easier to digest this if you could provide a minimal example that demonstrates the essence of your problem http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports.

A few scattered thoughts

1. You probably don't want to mix dask.delayed and dask.array or dask.dataframe. Something like `dask.delayed(np.all)(dask_array)` is strange. See http://docs.dask.org/en/latest/best-practices.html and http://docs.dask.org/en/latest/delayed-best-practices.html#don-t-call-dask-delayed-on-other-dask-collections
2. I'm not familiar with `np.select`, but it seems somewhat similar to `np.where`. Perhaps that would suffice for you, or perhaps `dask.array.select` could be implemented (separate issue?)
3. > There is no easy way to initialize dataframe by columns

Correct. Dask datafrmaes are partitioned along the index. This seems to fit your problem structure well (longer than wider). Why was this an issue? 

A simple, focused example would make it easier for us to provide guidance / find what needs to be done here.

**Comment by hoangthienan95** [comment]
Hi @TomAugspurger, I'll try but it's hard to provide a minimal example to pull out the main piece of this: the delayed object returned by another package's data parser (who uses `dask`). I'll try my best to respond clearly to your points.
1. the `test_arr` is not a dask array, but it's actually a `delayed` object that returns a `numpy array` when computed. In order to use `np.all` to get boolean from comparison, I have to either compute `test_arr` or wrap `np.all` with delayed, which I did.
1. Yes the use of `np.select` is for convenience, when you have multiple conditions mapped to multiple values
1. I need to map the list of `test_arr`objects into columns of a parquet file since that would help my later steps in analysis pipeline.

> I need each each test_arr to map to a column because in next step of the pipeline, I only need a subset of columns as any given moment (but all the columns should be batched together as a dataset entity).

Basically, I cant find a way to put `divisions` information into my delayed `np.select`/`np.where` object so that it can be appended as columns into my base dataframe below without triggering computation. I'm trying to keep it as DAG until the `to_parquet` step.

![image](https://user-images.githubusercontent.com/25307953/57731357-77cb4f80-7668-11e9-979e-a5a2044cd205.png)

Let me know if this clarifies!

**Comment by hoangthienan95** [comment]
The data parser gives `genotype`, which is a list of delayed dictionaries. The array I care about is in the key `"probs"`
![image](https://user-images.githubusercontent.com/25307953/57731732-59198880-7669-11e9-9f72-711dac66e6ec.png)

![image](https://user-images.githubusercontent.com/25307953/57731675-37b89c80-7669-11e9-8f30-a8d817912af3.png)

**Comment by hoangthienan95** [comment]
I try to keep things delayed until `to_parquet` because I assume that parquet uses column store, so that each column can be computed in sequence and not all at once?

**Comment by TomAugspurger** [comment]
I'll certainly acknowledge that providing minimal examples is difficult :)
It seems like my initial skim was off base, so I'll hope that you can
get something put together.

As an aside, code snippets are typically preferred to screenshots (makes it
easier to try out & think about things locally).

**Comment by TomAugspurger** [comment]
@hoangthienan95 how are things going here? Have you had time to produce a minimal example demonstrating the problem(s)?

**Comment by mrocklin** [comment]
Closing due to lack of response.  @hoangthienan95 if you're able to provide a minimal example please respond here and we'll reopen.

**Comment by hoangthienan95** [comment]
Hi there, sorry for being slow to get back to you, the past few weeks have been busy.

One of the problems I am facing is doing operations/assign new columns from current columns where the index is not monotonically increasing and doesn't start from 0. My `id` index is the result from earlier merging operations and I would like to keep it for future operations.

I'm having a hard time creating new column while keeping the index and so there's misalignment when assigning the new column:

```python
test_df = pd.DataFrame({'case_control_status': {1000194: 1,
  1000783: 0,
  1001009: 0,
  1001289: 0,
  1001347: 0},
 'sex_str': {1000194: 'Female',
  1000783: 'Female',
  1001009: 'Female',
  1001289: 'Male',
  1001347: 'Male'}}
)

test_ddf = dd.from_pandas(test_df, npartitions = 2)

```

In pandas, this works:
test_df.assign(sex = np.where(test_df["sex_str"] == "Male",1,0))

In dask, from what I know, the only way to assign a column is to turn it into a Dask Series? Here are the approaches I tried:

```python
test_ddf.eval("sex = 1 if sex_str == 'Male' else 0")  #NotImplementedError: 'IfExp' nodes are not implemented

test_ddf["sex_str"].map_partitions(lambda x: np.where(x == "Male", 1, 0)) #returns dask array with no index and can’t assign

test_ddf.map_partitions(lambda x: np.where(x.sex_str == "Male", 1,0)) #same as above

dd.from_dask_array(test_ddf["sex_str"].map_partitions(lambda x: np.where(x == "Male", 1 , 0))) doesn't keep index

dd.from_dask_array(da.where(test_ddf["sex_str"] == "Male", 1, 0)) #index error, tuple index out of range

dd.from_array(np.where(test_ddf["sex_str"] == "Male", 1, 0)) #doesn't keep the index
```

The only thing that worked was:

```python
test_ddf["sex_str"].map_partitions(lambda x: x.apply(lambda y: 1 if y == "Male" else 0))
``` 

but it seems unnecessarily complicated. Basically, I need a way to assign the newly created Series the same index as `test_ddf` or reset the index of `test_ddf` to monotonically increasing (which I tried with `test_ddf.reset_index().reset_index().set_index("index"). Am I missing something?

I looked at:
https://github.com/dask/dask/issues/2690#issuecomment-427838196
https://github.com/dask/dask/issues/1426

but still don't know how to proceed

**Comment by TomAugspurger** [comment]
On the narrow question of how to add a column, in this case I would recommend

```python
In [17]: test_ddf.assign(sex=(test_ddf.sex_str == 'Male').astype(int))
Out[17]:
Dask DataFrame Structure:
              case_control_status sex_str    sex
npartitions=2
1000194                     int64  object  int64
1001289                       ...     ...    ...
1001347                       ...     ...    ...
Dask Name: assign, 10 tasks

```

**Comment by hoangthienan95** [comment]
Thank you @TomAugspurger, that worked. However, I have other parts of my code where I have to add columns while keeping the index that this cannot be applied. In general, if you can create an array from applying a numpy function on dask dataframe `df`, how do you convert this to a Dask Series while retaining the index of `df` so it could be appended? (assuming that the function is also implemented in `dask.array`)

**Comment by TomAugspurger** [comment]
I think something like `dd.from_dask_array(dask_array, index=df.index)`
would work.
